### PR TITLE
[NOD-1067] Add deletion to ffdb

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

**Comment by svarogg** [comment]
Please create big branch for all pruning stage 3 and update this PR to go to it.

File: database/ffldb/ff/flatfiledb.go [hunk 1]
```diff
@@ -101,3 +103,29 @@ func (ffdb *FlatFileDB) store(storeName string) (*flatFileStore, error) {
 	}
 	return store, nil
 }
+
+// DeleteUpToLocation deletes the store data up to the given location file.
```
Comment (svarogg) [inline database/ffldb/ff/flatfiledb.go:107]: Please update comment to something similar to what's written in the interface, explaining about the guarantees this gives

File: database/location.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+// Serialize returns a byte slice that represents the StoreLocation
+func (s StoreLocation) Serialize() []byte {
+	serializedLocation := make([]byte, len(s))
+	copy(serializedLocation, s)
+	return serializedLocation
+}
+
+// Deserialize deserializes the given byte slice into s.
+func (s *StoreLocation) Deserialize(serializedLocation []byte) {
```
Comment (svarogg) [inline database/location.go:14]: Why not make Deserialize a package-level function that returns StoreLocation? (ofc, rename it to DeserializeStoreLocation or something similar)

Comment (svarogg) [reply database/location.go:14]: Also: I think we want to check that serializedLocation is a valid representation of StoreLocation

File: database/ffldb/ff/delete.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			return errors.Wrapf(err, "failed to delete file "+
+				"number %d in store '%s'", fileNumber, s.storeName)
+		}
+	}
+
+	return nil
+}
+
+func (s *flatFileStore) fileExists(fileNumber uint32) (bool, error) {
+	filePath := flatFilePath(s.basePath, s.storeName, fileNumber)
```
Comment (svarogg) [inline database/ffldb/ff/delete.go:71]: Here also, check for s.isClosed.

File: database/database.go [hunk 4]
```diff
@@ -11,6 +11,12 @@ package database
 type Database interface {
 	DataAccessor
 
+	// DeleteUpToLocation deletes as much data as it can from the given store, while guaranteeing
+	// that the data belongs to dbLocation, its following locations and dbPreservedLocations will
+	// be kept.
```
Comment (stasatdaglabs) [inline database/database.go:16]: Please rephrase it to "while keeping the data that dbLocation, its following locations, and dbPreservedLocations point to."

File: database/ffldb/ff/delete_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+		t.Fatalf("store.deleteUpToFile(): %s", err)
+	}
+
+	for fileNumber := range preservedFiles {
+		exists, err := store.fileExists(fileNumber)
+		if err != nil {
+			t.Fatalf("store.fileExists(): %s", err)
+		}
+		if !exists {
+			t.Errorf("file %d is inpreservedFiles  was expected to be preserved", fileNumber)
```
Comment (stasatdaglabs) [inline database/ffldb/ff/delete_test.go:46]: Please move one of the spaces after "inpreservedFiles" to after the "in".

File: database/database.go (hunk 4 above)
Comment (stasatdaglabs) [reply database/database.go:16]: Also, make sure to propagate this change to everywhere else this comment is written.
